pygmsh/built_in/geometry.py

```diff
@@ -255,9 +255,10 @@ def add_circle(
                 ])
 
         # Apply the transformation.
-        # TODO assert that the transformation preserves circles
         if R is not None:
-            X = [numpy.dot(R, x) + x0 for x in X]
+            # Assert that the transformation preserves circles
+            assert numpy.allclose(abs(numpy.linalg.eigvals(R)), 1.)
```

Comment:
An error message here would be nice, e.g.,
```
assert [...], 'The transformation matrix does not preserve circles. (At least one eigenvalue != +-1.)'
```

Reply:
Good idea.

I think some valid `R` have complex eigenvalues, but always with unit modulus; e.g. `pygmsh.rotation_matrix([1, 0, 0], pi/4)` has +1 and the complex conjugate pair (1± j) / √2.

I had wondered whether only the spectrum relative to the _xy_-plane subspace mattered, but saw that `R` is passed out as an attribute of the Circle and wasn't sure where else that might be used so left test as is.